### DISCUSSION TITLE
Error when sending test mail (using gmail)

### DISCUSSION THREAD

**Discussion by maeries** [dani-garcia/vaultwarden#1360] [Q&A] [answered]
I want to setup bitwarden to be able to send mails. For the mail service i would like to use gmail (or does bitwarden come with its own?)
This is my (alienated) configuration:
![config](https://user-images.githubusercontent.com/30597531/77368375-62c94d80-6d5c-11ea-9133-f162310aa297.png)
Now, when I try to send a test mail I get a popup saying: Error sending SMTP test email LetreErr

The shell says:

`bitwarden_1  | [2020-03-23 22:22:17][request][INFO] POST /admin/test/smtp/`
`bitwarden_1  | [2020-03-23 22:22:27][error][ERROR] LetreErr.`
`bitwarden_1  | [CAUSE] Parsing(`
`bitwarden_1  |     Tag,`
`bitwarden_1  | )`
`bitwarden_1  | [2020-03-23 22:22:27][response][INFO] POST /admin/test/smtp (test_smtp) => 400 Bad Request`

I can't find any logfile though.

I tried it with and without explicit TLS and also with the ports 465 and 587.

I run bitwarden 1.14.1-91dd1947 using docker-compose on a ubuntu LXC.

**Comment by jjlin** [comment]
Try it with `Enable SSL` checked, `Use explicit TLS` unchecked, and port 587. Also, if you have 2FA enabled on your Google account, you need to use an app password instead of your normal Google password (see https://support.google.com/mail/answer/185833).

**Comment by tnds82** [comment]
The same problem here.
Configuration was performed with Enable SSL checked, Use explicit TLS unchecked, and port 587
I tested with two different gmail accounts one with 2FA and one without 2FA.
I created the application password and it continues to give the same error

**Comment by maeries** [comment]
With explicit TLS and Port 587 I get the same message. The difference is that it now appears instantly instead of being delayed 10s. I actually used a stop watch, so I'm sure it's not the 15s timeout that caused it before.
The account I'm using does not have 2FA

**Comment by jjlin** [comment]
It works for me with the latest Docker image and the configuration I mentioned (I use 2FA). Perhaps your network provider blocks port 587? Try `telnet smtp.gmail.com 587` and see if you get a response like `220 smtp.gmail.com ESMTP k20sm783595pgn.62 - gsmtp`.

**Comment by maeries** [comment]
The port isn't blocked. I get the response you mentioned

**Comment by VolkanoLiu** [comment]
Maybe you can `apt install libperl5.28` in the container. I solved this problem by the method. I found this problem when I try another Perl script.

**Comment by maeries** [comment]
I can't get that to work unfortunately 

`root@f83766827d96:/# apt install libperl5.28`
`Reading package lists... Done`
`Building dependency tree       `
`Reading state information... Done`
`E: Unable to locate package libperl5.28`
`E: Couldn't find any package by glob 'libperl5.28'`
`E: Couldn't find any package by regex 'libperl5.28'`

**Comment by VolkanoLiu** [comment]
> I can't get that to work unfortunately
> 
> `root@f83766827d96:/# apt install libperl5.28`
> `Reading package lists... Done`
> `Building dependency tree `
> `Reading state information... Done`
> `E: Unable to locate package libperl5.28`
> `E: Couldn't find any package by glob 'libperl5.28'`
> `E: Couldn't find any package by regex 'libperl5.28'`

You should firstly `apt update`.
The libperl version is not always the same, so you can use `apt search libperl5` to find the proper version and then install it.
By the way, if the container's MTU is not right, the update will be very slow.

Edit: install libperl cannot solve the problem, just change the MTU.

**Comment by jjlin** [comment]
I highly doubt perl has anything to do with this, given bitwarden_rs doesn't call out to perl to send mail. However, you might try using something like https://github.com/mludvig/smtp-cli from within your bitwarden_rs container to see if you can gain any insight into what's going wrong. Note that you will need perl and various modules in the container for this tool; see its README.md for details.

**Comment by VolkanoLiu** [comment]
> I highly doubt perl has anything to do with this, given bitwarden_rs doesn't call out to perl to send mail. However, you might try using something like https://github.com/mludvig/smtp-cli from within your bitwarden_rs container to see if you can gain any insight into what's going wrong. Note that you will need perl and various modules in the container for this tool; see its README.md for details.

Yes, I'm sorry I made a mistake.
I think the wrong MTU is the real reason.
I remove the libperl and the SMTP works fine. But when I reset the MTU to the default value (1500), the LetreErr happens again. It's better to set the MTU the same as physical network adapter.
I change the MTU and install libperl at the same time, so I made the mistake.

To change the MTU, this is a way:
```
networks:
default:
  driver: bridge
  driver_opts:
    com.docker.network.driver.mtu: 1500
```

**Comment by maeries** [comment]
> ```
> networks:
> default:
>   driver: bridge
>   driver_opts:
>     com.docker.network.driver.mtu: 1500
> ```

I guess I have to put that in the docker-compose.yml, right?

But what do I set the MTU to? Both the ubuntu LXC container and Proxmox show an MTU of 1500 using `ip a` and also `apt update` in the docker container seemed to be at normal speed. So I guess 1500 is correct

**Comment by maeries** [comment]
As the guy in #943, I also used a gmail accont with a custom domain, but unlike them I couldn't solve it by using another account. I tried a different google account and one from a different provider. The error is still the same

Edit: I just noticed that this time I got a different output in the console. I also got this one when I first encountered this problem, but I have no idea why I got the one from my first post temporarily. 

```
bitwarden_1  | [2020-04-01 16:34:36][request][INFO] POST /admin/test/smtp/
bitwarden_1  | [2020-04-01 16:34:36][error][ERROR] LetreErr.
bitwarden_1  | [CAUSE] Io(
bitwarden_1  |     Custom {
bitwarden_1  |         kind: Other,
bitwarden_1  |         error: Failure(
bitwarden_1  |             Ssl(
bitwarden_1  |                 Error {
bitwarden_1  |                     code: ErrorCode(
bitwarden_1  |                         1,
bitwarden_1  |                     ),
bitwarden_1  |                     cause: Some(
bitwarden_1  |                         Ssl(
bitwarden_1  |                             ErrorStack(
bitwarden_1  |                                 [
bitwarden_1  |                                     Error {
bitwarden_1  |                                         code: 336130315,
bitwarden_1  |                                         library: "SSL routines",
bitwarden_1  |                                         function: "ssl3_get_record",
bitwarden_1  |                                         reason: "wrong version number",
bitwarden_1  |                                         file: "../ssl/record/ssl3_record.c",
bitwarden_1  |                                         line: 332,
bitwarden_1  |                                     },
bitwarden_1  |                                 ],
bitwarden_1  |                             ),
bitwarden_1  |                         ),
bitwarden_1  |                     ),
bitwarden_1  |                 },
bitwarden_1  |                 X509VerifyResult {
bitwarden_1  |                     code: 0,
bitwarden_1  |                     error: "ok",
bitwarden_1  |                 },
bitwarden_1  |             ),
bitwarden_1  |         ),
bitwarden_1  |     },
bitwarden_1  | )
```

**Comment by jjlin** [comment]
Did you click the save button to save your settings before testing each configuration? Otherwise the send test email functionality won't pick up on the changes.

**Comment by maeries** [comment]
Yes, i did

**Comment by kassyss** [comment]
Hello,
Until today i was using smtp with port 25 (and no authentication) but because i wanted to be "more secure" i have played with Bitwarden_rs environment file to setup mail with my own selfhosted mailserver using port 587.

The following configuration is not working
```
SMTP_HOST=smtp.domain.com
SMTP_FROM=bitwarden-rs@domain.com
SMTP_FROM_NAME=Bitwarden_RS
SMTP_PORT=587
SMTP_SSL=true
#SMTP_EXPLICIT_TLS=true
SMTP_USERNAME=mailbox-username@domain.com
SMTP_PASSWORD=hxxxxxxxxxxxxxxxxx
SMTP_TIMEOUT=15
```

The following configuration is working

```
SMTP_HOST=smtp.domain.com
SMTP_FROM=mailbox-username@domain.com
SMTP_FROM_NAME=Bitwarden_RS
SMTP_PORT=587
SMTP_SSL=true
#SMTP_EXPLICIT_TLS=true
SMTP_USERNAME=mailbox-username@domain.com
SMTP_PASSWORD=hxxxxxxxxxxxxxxxxx
SMTP_TIMEOUT=15
``` 
So i conclude that there is a problem if SMTP_FROM differs from SMTP_USERNAME
(My mail server allow to change the MAIL FROM)

Hope this can help.

Best regards,

**Comment by maeries** [comment]
unfortunately I already had set SMTP_FROM and SMTP_USERNAME to the same value, so that is not the reason for this issue

**Comment by yzhblind** [comment]
I got the problem in a different way.
When I tried to send a test mail, I also got "Error sending SMTP test email LetreErr".

`[2020-04-05 11:53:19][start][INFO] Rocket has launched from http://0.0.0.0:80`
`[2020-04-05 12:00:25][request][INFO] GET /admin/`
`[2020-04-05 12:00:25][response][INFO] GET /admin (admin_page) => 200 OK`
`[2020-04-05 12:00:39][request][INFO] POST /admin/test/smtp/`
`[2020-04-05 12:01:09][error][ERROR] LetreErr.`
`[CAUSE] Io(`
    `Os {`
        `code: 11,`
        `kind: WouldBlock,`
        `message: "Resource temporarily unavailable",`
   ` },`
`)`
`[2020-04-05 12:01:09][response][INFO] POST /admin/test/smtp (test_smtp) => 400 Bad Request`

**Comment by kassyss** [comment]
Have you tried to test connectivity and dns resolution (smtp gmail) From the Bitwarden_rs container itself (docker exec -it container-name sh)?

**Comment by maeries** [comment]
looks fine

```
root@874fb49318f0:/# host smtp.gmail.com
smtp.gmail.com has address 173.194.76.109
smtp.gmail.com has IPv6 address 2a00:1450:400c:c00::6d
```
Edit:
interestingly I get another result if I run the command on my desktop PC

```
host smtp.gmail.com
smtp.gmail.com has address 74.125.140.109
smtp.gmail.com has IPv6 address 2a00:1450:400c:c09::6d
```
Edit2:
I set the DNS server of the container to 1.1.1.1 an now I get the same result as on my PC. Does not change anything regarding the problem though

**Comment by jonathanmmm** [comment]
Same here, have change nothing and suddenly it does not work anymore.

I get the same error on the webinterface but different in log:
[2020-04-09 08:12:03][request][INFO] POST /admin/test/smtp/
[2020-04-09 08:12:19][error][ERROR] LetreErr.
[CAUSE] Io(
    Custom {
        kind: TimedOut,
        error: "connection timed out",
    },
)
[2020-04-09 08:12:19][response][INFO] POST /admin/test/smtp (test_smtp) => 400 Bad Request

I run the bitwardenrs/server:raspberry docker on Rasbian stretch.

Strangely I couldn't change settings and send a test email through logging in through my nginx proxy. Also there were no profile icons for the user, I believe. I have connected directly with the rocket port that goes out of docker. There it works (means I can change settings and request to send test email, but it fails, as the error log shows).

**Comment by maeries** [comment]
UPDATE: I upgraded to 1.14.2-5a390a97, but the error still occurs

**Comment by jjlin** [comment]
Nothing has changed in the email sending code recently, so I wouldn't expect an upgrade to make a difference. Did you try my suggestion of seeing if you can get it work using another tool like `smtp-cli`?

Also, unless you're dead set on using Gmail for some reason, you might try something like [SendGrid](https://sendgrid.com/) or [MailJet](https://www.mailjet.com/) instead. These all offer 100-200 emails per day on the free tier.

**Comment by jonathanmmm** [comment]
I fixed it only after I changed the provider. I had two E-Mail Addresses from the same provider, both didn't work. So I changed to another. Even as both providers actually  belong to the same big company it worked with the other provider. Don't know why, this time he could reach the smtp server (a diffferent dns name, as the first provider) and everything worked.

Had already the problem that the Provider has an automatic system, as if you normally send 3 E-Mails a day and you connect it to a logging service and you send like 100 E-Mails a day that you can get blocked to send any E-Mail, more than 3 or so. This is spam protection. Switching to the free trial of the premium plan and back resetted this timer. You have to train it that you send more E-Mails. This is probably not important, as Bitwarden does send very few E-Mails, wanted still to give my experience.

**Comment by maeries** [comment]
You both are right. The provider was the problem. I already had tested autistici.org which also caused the LetreErr. Now I'm using hosteurope.de with SSL, without explicit TLS and port 587. That works

**Comment by floatingpurr** [comment]
Same problem here with the SMTP config for a Gmail account. I cannot figure out how to use Gmail without enabling less secure apps...

**Comment by ghost** [comment]
i have less secure apps and doesn't works. I've tried all the settings mentioned here.

Anyone that has it working can post the configuration used for gmail ?

Thanks in advance

**Comment by ScarlitoDoBrazil** [comment, marked as answer]
> 
> 
> i have less secure apps and doesn't works. I've tried all the settings mentioned here.
> 
> Anyone that has it working can post the configuration used for gmail ?
> 
> Thanks in advance

Hi ppadial.
I finaly got it working ! Here is the configuration I use :

Enabled : **true**
Host : **smtp.gmail.com**
Enable SSL : **true**
Use explicit TLS : **false**
Port : **587**
From Address : **<my_gmail_address@gmail.com>**
From Name : **< The Name I want to show as Sender >**
Password : **< the "less secure app" password >**
Json form auth mechanism : **empty**
SMTP connection timeout : **15**
Server name sent during HELO : **empty**

I **_saved_** the settings before testing.
Hope this can help to make it work for you and others.

Regards.

**Comment by ghost** [comment]
Thank you very much, i got it working too!!

**Comment by ghost** [comment]
> > i have less secure apps and doesn't works. I've tried all the settings mentioned here.
> > Anyone that has it working can post the configuration used for gmail ?
> > Thanks in advance
> 
> Hi ppadial.
> I finaly got it working ! Here is the configuration I use :
> 
> Enabled : **true**
> Host : **smtp.gmail.com**
> Enable SSL : **true**
> Use explicit TLS : **false**
> Port : **587**
> From Address : **[my_gmail_address@gmail.com](mailto:my_gmail_address@gmail.com)**
> From Name : **< The Name I want to show as Sender >**
> Password : **< the "less secure app" password >**
> Json form auth mechanism : **empty**
> SMTP connection timeout : **15**
> Server name sent during HELO : **empty**
> 
> I **_saved_** the settings before testing.
> Hope this can help to make it work for you and others.
> 
> Regards.

YOU ARE A GENIUS! THIS WORKED!
